### PR TITLE
Add int4 example for convnd_fwd_bias_relu_add

### PR DESCRIPTION
**Pull request by geyyer**
- Add an example for int4
- Refactor common header and other examples to allow *UserDataType besides *KernelDataType
- Update CMakeLists.txt
- Add int4 implementation of AddReluAdd to element_wise_operation.hpp
- Add int4 implementation in check_err.hpp

### CODE REVIEW

File: example/30_grouped_convnd_fwd_bias_relu_add/CMakeLists.txt [hunk 1]
```diff
@@ -8,4 +8,9 @@ add_example_executable(example_grouped_convnd_fwd_bias_relu_add_xdl_bf16 grouped
 target_link_libraries(example_grouped_convnd_fwd_bias_relu_add_xdl_bf16 PRIVATE utility)
 
 add_example_executable(example_grouped_convnd_fwd_bias_relu_add_xdl_int8 grouped_convnd_fwd_bias_relu_add_xdl_int8.cpp)
-target_link_libraries(example_grouped_convnd_fwd_bias_relu_add_xdl_int8 PRIVATE utility)
\ No newline at end of file
+target_link_libraries(example_grouped_convnd_fwd_bias_relu_add_xdl_int8 PRIVATE utility)
+
+if(USE_BITINT_EXTENSION_INT4)
+  add_example_executable(example_grouped_convnd_fwd_bias_relu_add_xdl_int4 grouped_convnd_fwd_bias_relu_add_xdl_int4.cpp)
+  target_link_libraries(example_grouped_convnd_fwd_bias_relu_add_xdl_int4 PRIVATE utility)
```
Comment (aosewski) [inline example/30_grouped_convnd_fwd_bias_relu_add/CMakeLists.txt:15]: You don't have to link `utility` library manually. It's already done in `add_example_executable`. Please do fix all the occurences in this CMakeLists file.

File: example/30_grouped_convnd_fwd_bias_relu_add/grouped_convnd_fwd_bias_relu_add_common.hpp [hunk 2]
```diff
@@ -64,21 +67,21 @@ int run_grouped_conv_fwd_bias_relu_add(bool do_verification,
...
-    DeviceMem in_device_buf(sizeof(InDataType) * in.mDesc.GetElementSpaceSize());
-    DeviceMem wei_device_buf(sizeof(WeiDataType) * wei.mDesc.GetElementSpaceSize());
-    DeviceMem bias_device_buf(sizeof(OutDataType) * bias.mDesc.GetElementSpaceSize());
-    DeviceMem residual_device_buf(sizeof(OutDataType) * residual.mDesc.GetElementSpaceSize());
-    DeviceMem out_device_buf(sizeof(OutDataType) * out_device.mDesc.GetElementSpaceSize());
+    DeviceMem in_device_buf(sizeof(InUserDataType) * in.mDesc.GetElementSpaceSize());
+    DeviceMem wei_device_buf(sizeof(WeiUserDataType) * wei.mDesc.GetElementSpaceSize());
+    DeviceMem bias_device_buf(sizeof(OutUserDataType) * bias.mDesc.GetElementSpaceSize());
+    DeviceMem residual_device_buf(sizeof(OutUserDataType) * residual.mDesc.GetElementSpaceSize());
+    DeviceMem out_device_buf(sizeof(OutUserDataType) * out_device.mDesc.GetElementSpaceSize());
```
Comment (aosewski) [inline example/30_grouped_convnd_fwd_bias_relu_add/grouped_convnd_fwd_bias_relu_add_common.hpp:84]: In order to get correct results with int4 data type, you have to convert data from int4 Tensor into int8 tensor and only then such converted Tensor can be passed as device op argument. Additionally results should also be converted back to int4 Tensor.

**Review by aosewski — CHANGES_REQUESTED** [review]
Please keep in mind the need for int4 tensor conversion before data transfer to device and converting results back after retrieving results.

File: example/30_grouped_convnd_fwd_bias_relu_add/CMakeLists.txt (hunk 1 above)
Comment (geyyer) [reply example/30_grouped_convnd_fwd_bias_relu_add/CMakeLists.txt]: Fixed in [439db12](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/375/commits/439db121bb7109de4d70ac6173824e0ce83823fa)

File: example/30_grouped_convnd_fwd_bias_relu_add/grouped_convnd_fwd_bias_relu_add_common.hpp (hunk 2 above)
Comment (geyyer) [reply example/30_grouped_convnd_fwd_bias_relu_add/grouped_convnd_fwd_bias_relu_add_common.hpp:84]: Fixed in [a223220](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/375/commits/a223220955d8720381a517caf58b6a76079aa5a1)

**Comment by asroy** [comment]
@geyyer CI failure

File: library/include/ck/library/utility/host_tensor.hpp [hunk 3]
```diff
@@ -271,6 +271,7 @@ struct Tensor
     ~Tensor() = default;
 
     Tensor& operator=(const Tensor&) = default;
+
```
Comment (aosewski) [inline library/include/ck/library/utility/host_tensor.hpp:274]: Let's not make this file unnecessarily "wider" :)

**Review by aosewski — APPROVED** [review]
As long as CI pass looks good :+1:

File: library/include/ck/library/utility/host_tensor.hpp (hunk 3 above)
Comment (geyyer) [reply library/include/ck/library/utility/host_tensor.hpp]: Oops, fixed in [a7da01f](https://github.com/ROCmSoftwarePlatform/composable_kernel/pull/375/commits/a7da01f0c12537fbf5df4c8d0b3cd8ee1b7f655e)

**Comment by geyyer** [comment]
CI fixed

**Review by aosewski — APPROVED** [review]

**Pull request merged**